// @splitsoftware/splitio-sync-tools/lib/cjs/settings/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.synchronizerSettingsValidator = void 0;
const constants_1 = require("@splitsoftware/splitio-commons/cjs/utils/constants");
const lang_1 = require("@splitsoftware/splitio-commons/cjs/utils/lang");
const index_1 = require("@splitsoftware/splitio-commons/cjs/utils/settingsValidation/index");
const builtinLogger_1 = require("@splitsoftware/splitio-commons/cjs/utils/settingsValidation/logger/builtinLogger");
const defaults_1 = require("./defaults");
const FLAG_SPEC_VERSIONS = ['1.0', constants_1.FLAG_SPEC_VERSION];
/**
 * Object with some default values to instantiate the application and fullfil internal
 * requirements.
 */
const params = {
    logger: builtinLogger_1.validateLogger,
    defaults: defaults_1.defaults,
    consent: () => undefined,
    runtime: () => { return { ip: false, hostname: false }; },
    flagSpec: ({ sync: { flagSpecVersion }, log }) => {
        if (FLAG_SPEC_VERSIONS.indexOf(flagSpecVersion) > -1)
            return flagSpecVersion;
        log.error(`settings: you passed an invalid "flagSpecVersion" config param. It should be one of the following values: ${FLAG_SPEC_VERSIONS.map((version => `"${version}"`))}. Defaulting to "${constants_1.FLAG_SPEC_VERSION}"`);
        return constants_1.FLAG_SPEC_VERSION;
    },
};
function validatePositiveInteger(log, paramName, actualValue, defaultValue) {
    if ((0, lang_1.isIntegerNumber)(actualValue) && actualValue > 0)
        return actualValue;
    log.warn(`'${paramName}' parameter must be a positive integer number. Using default value (${defaultValue}) instead.`);
    return defaultValue;
}
/**
 * Function to validate Synchronizer config.
 *
 * @param {any} config  Synchronizer config object provided by the user.
 * @returns {ISettings}
 */
function synchronizerSettingsValidator(config) {
    // @ts-ignore
    const settings = (0, index_1.settingsValidation)(config, params);
    // @ts-ignore, override readonly prop
    settings.mode = undefined; // "producer" mode
    const { scheduler, log } = settings;
    // @TODO validate synchronizerMode eventually
    // @TODO: validate minimum and maximum value for config params.
    scheduler.eventsPerPost = validatePositiveInteger(log, 'eventsPerPost', scheduler.eventsPerPost, defaults_1.defaults.scheduler.eventsPerPost);
    scheduler.impressionsPerPost = validatePositiveInteger(log, 'impressionsPerPost', scheduler.impressionsPerPost, defaults_1.defaults.scheduler.impressionsPerPost);
    scheduler.maxRetries = validatePositiveInteger(log, 'maxRetries', scheduler.maxRetries, defaults_1.defaults.scheduler.maxRetries);
    return settings;
}
exports.synchronizerSettingsValidator = synchronizerSettingsValidator;
